tests/test_userspace.py

```diff
@@ -87,3 +91,64 @@ def test_hostname(shell, check):
         assert serial_number in etc_hostname
 
 
+@pytest.fixture
+def clocktree(shell):
+    """
+    Read the clock tree from the DUT and parse it into a data structure.
+    """
+    re_entry = re.compile(r"^\s*(\S+)\s+\d+\s+\d+\s+\d+\s+(\d+)\s+\d+\s+\d+\s+(\d+)\s+\S\s+(\S+)\s+")
+    re_2nd = re.compile(r"^\s+(\S+)\s+\S+\s+$")
+
+    @dataclass
+    class Clk:
+        clk_name: str
+        rate: int
+        duty: int
+        consumer: list
+
+    clks = {}
+    clk = None
+    for line in shell.run_check("cat /sys/kernel/debug/clk/clk_summary"):
+        if match := re_entry.match(line):
+            if clk:
+                clks[clk.clk_name] = clk
+            clk = Clk(clk_name=match.group(1), rate=int(match.group(2)), duty=int(match.group(3)), consumer=[])
+            if match.group(4) != "deviceless":
+                clk.consumer.append(match.group(4))
+            continue
+
+        if match := re_2nd.match(line):
+            if match.group(1) != "deviceless":
+                clk.consumer.append(match.group(1))
+
+    return clks
+
+
+@pytest.mark.parametrize(
+    "clock_name, rate, consumer",
+    (
+        # Ethernet Clocks: Needed for the communication with the phy to work
+        ("ethptp_k", 125000000, ("5800a000.ethernet",)),
+        ("ethck_k", 125000000, ("5800a000.ethernet",)),
+        ("ethrx", 125000000, ("5800a000.ethernet",)),
+
+        # CAN Clock: Chosen to be 48MHz for minimum baudrate error across all rates
+        ("fdcan_k", 48000000, ("4400f000.can", "4400e000.can")),
+    )
+)
+def test_clocktree(clocktree, check, clock_name, rate, consumer):
```

Comment:
Will the `clocktree` fixture be set up per `parametrize`d call or once for all of them?
If it is the latter, should we maybe mark it with `module` [scope](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session)?
That would make make it very cheap to maybe add even more clocks to check for. For example USB and serial (to make sure we can generate all relevant baud rates).

Reply:
Reading the clocktree over UART is actually expensive. Simply because it's huge.
The problem here is, that the `shell`-fixture is `function`-scoped. So this fixture has to bee `function`-scoped, too.

I thought about working around this (by using some kind of cache), but I decided to go for the simple solution and postpone that for when we actually want to check more clocks.

Reply:
Okay. I guess this is okay then.